src/kudu/consensus/consensus_queue.cc

```diff
@@ -883,6 +887,118 @@ void PeerMessageQueue::AdvanceQueueWatermark(const char* type,
   }
 }
 
+void PeerMessageQueue::AdvanceMajorityReplicatedWatermark(
+    int64_t* watermark, const OpId& replicated_before,
+    const OpId& replicated_after,
+    const std::map<std::string, int>& data_commit_quorum,
+    ReplicaTypes replica_types, const TrackedPeer* who_caused) {
+  CHECK(watermark);
+  CHECK(who_caused);
+
+  if (VLOG_IS_ON(2)) {
+    VLOG_WITH_PREFIX_UNLOCKED(2) << "Updating majority_replicated watermark: "
+        << "Peer (" << who_caused->ToString() << ") changed from "
+        << replicated_before << " to " << replicated_after << ". "
+                                 << "Current value: " << *watermark;
+  }
+
+  // For each region, we compute a vector of indexes that were replicated.
+  // It might look like the following example:
+  // prn: <4,4,5,7>
+  // frc: <2,3,4>
+  // lla: <5,5>
+  // This example suggests that we received non-erroneous responses from 4
+  // replicas in prn, 3 in frc and 2 in lla. Two replicas in prn have received
+  // entries until index 4, one has received until 5 and one until 7. Similarly,
+  // 2 replicas in lla have received entries until index 5.
+  std::map<std::string, vector<int64_t> > watermarks_by_region;
+  for (const PeersMap::value_type& peer : peers_map_) {
+    if (replica_types == VOTER_REPLICAS &&
+        peer.second->peer_pb.member_type() != RaftPeerPB::VOTER) {
+      continue;
+    }
+    // Refer to the comment in AdvanceQueueWatermark method for why only
+    // successful last exchanges are considered.
+    if (peer.second->last_exchange_status == PeerStatus::OK) {
+      const string& peer_region = peer.second->peer_pb.attrs().region();
+      std::vector<int64_t>& regional_watermarks = LookupOrInsert(
+          &watermarks_by_region, peer_region, std::vector<int64_t>());
+      regional_watermarks.push_back(peer.second->last_received.index());
+    }
+  }
+
+  // Vector of maximum indexes which satisfy commit requirements in each region.
+  std::vector<int64_t> commit_index_regional_level;
+  for (const std::pair<std::string, int>& regional_commit_requirement :
+      data_commit_quorum) {
+    std::map<std::string, std::vector<int64_t> >::iterator it =
+        watermarks_by_region.find(regional_commit_requirement.first);
+
+    // Return without advancing the commit watermark, if any region specified in
+    // the data commit quorum is not satisfied, ie. not enough number of
+    // replicas have responded from that region.
+    if (it == watermarks_by_region.end() ||
+        it->second.size() < regional_commit_requirement.second) {
+      VLOG_WITH_PREFIX_UNLOCKED(3)
+          << "Watermarks size: "
+          << ((it == watermarks_by_region.end()) ? 0 : it->second.size())
+          << ", Num peers required: " << regional_commit_requirement.second
+          << ", Region: " << regional_commit_requirement.first;
+      return;
+    }
+    std::sort(it->second.begin(), it->second.end());
```

Comment:
nit: If we sort these vectors right after we construct `watermarks_by_region` (say at L929), we don't have the cognitive overhead of wondering if/when we will sort them when reading this code later on. It becomes an invariant.